File: clients/Cpp/model/ColoredVertex.cpp
#include "ColoredVertex.hpp"

ColoredVertex::ColoredVertex() { }
ColoredVertex::ColoredVertex(std::shared_ptr<Vec2Float> worldPos, Vec2Float screenOffset, Color color) : worldPos(worldPos), screenOffset(screenOffset), color(color) { }
ColoredVertex ColoredVertex::readFrom(InputStream& stream) {
    ColoredVertex result;
    if (stream.readBool()) {
        result.worldPos = std::shared_ptr<Vec2Float>(new Vec2Float());
        *result.worldPos = Vec2Float::readFrom(stream);
    } else {
        result.worldPos = std::shared_ptr<Vec2Float>();
    }
    result.screenOffset = Vec2Float::readFrom(stream);
    result.color = Color::readFrom(stream);
    return result;
}
void ColoredVertex::writeTo(OutputStream& stream) const {
    if (worldPos) {
        stream.write(true);
        (*worldPos).writeTo(stream);
    } else {
        stream.write(false);
    }
    screenOffset.writeTo(stream);
    color.writeTo(stream);
}

File: clients/Ruby/model/colored_vertex.rb
require_relative 'vec2_float'
require_relative 'vec2_float'
require_relative 'color'
class ColoredVertex
    attr_accessor :world_pos
    attr_accessor :screen_offset
    attr_accessor :color
    def initialize(world_pos, screen_offset, color)
        @world_pos = world_pos
        @screen_offset = screen_offset
        @color = color
    end
    def self.read_from(stream)
        if stream.read_bool()
            world_pos = Vec2Float.read_from(stream)
        else
            world_pos = nil
        end
        screen_offset = Vec2Float.read_from(stream)
        color = Color.read_from(stream)
        ColoredVertex.new(world_pos, screen_offset, color)
    end
    def write_to(stream)
        if @world_pos.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            @world_pos.write_to(stream)
        end
        @screen_offset.write_to(stream)
        @color.write_to(stream)
    end
end
